src/java.base/share/classes/jdk/internal/foreign/abi/NativeEntryPoint.java

```diff
@@ -56,26 +56,38 @@ private NativeEntryPoint(MethodType methodType, long downcallStubAddress) {
 
     public static NativeEntryPoint make(ABIDescriptor abi,
                                         VMStorage[] argMoves, VMStorage[] returnMoves,
-                                        MethodType methodType, boolean needsReturnBuffer) {
+                                        MethodType methodType,
+                                        boolean needsReturnBuffer,
+                                        int capturedStateMask) {
         if (returnMoves.length > 1 != needsReturnBuffer) {
             throw new IllegalArgumentException("Multiple register return, but needsReturnBuffer was false");
         }
+        checkType(methodType, needsReturnBuffer, capturedStateMask);
 
-        assert (methodType.parameterType(0) == long.class) : "Address expected";
-        assert (!needsReturnBuffer || methodType.parameterType(1) == long.class) : "return buffer address expected";
-
-        CacheKey key = new CacheKey(methodType, abi, Arrays.asList(argMoves), Arrays.asList(returnMoves), needsReturnBuffer);
+        CacheKey key = new CacheKey(methodType, abi, Arrays.asList(argMoves), Arrays.asList(returnMoves), needsReturnBuffer, capturedStateMask);
         return NEP_CACHE.get(key, k -> {
-            long downcallStub = makeDowncallStub(methodType, abi, argMoves, returnMoves, needsReturnBuffer);
+            long downcallStub = makeDowncallStub(methodType, abi, argMoves, returnMoves, needsReturnBuffer, capturedStateMask);
             NativeEntryPoint nep = new NativeEntryPoint(methodType, downcallStub);
             CLEANER.register(nep, () -> freeDowncallStub(downcallStub));
             return nep;
         });
     }
 
+    private static void checkType(MethodType methodType, boolean needsReturnBuffer, int savedValueMask) {
+        if (methodType.parameterType(0) != long.class) {
+            throw new IllegalArgumentException("Address expected as first param: " + methodType);
```

Comment:
Is throwing IAE correct here? E.g. can the user do anything about it, or does the exception describe more of an internal error? (In that case AssertionError might be better?)

Reply:
Yes, it's an internal error. I can change the exception type